File: G_Projects/I_Vending_Machine/VendingMachineStates/VendingMachineContext.java
package G_Projects.I_Vending_Machine.VendingMachineStates;

import java.util.ArrayList;
import java.util.List;

import G_Projects.I_Vending_Machine.CommonEnums.Coin;
import G_Projects.I_Vending_Machine.PaymentStrategy.PaymentStrategy;
import G_Projects.I_Vending_Machine.UtilityClasses.Inventory;
import G_Projects.I_Vending_Machine.VendingMachineStates.ConcreteStates.IdleState;

public class VendingMachineContext {
    
    private VendingMachineState currMachineState; //Context owns the state
    private PaymentStrategy paymentStrategy;
    private Inventory inventory;
    private List<Coin> coinList;
    private int selectedItemCode;

    public VendingMachineContext(){
        this.currMachineState = new IdleState();
        this.inventory = new Inventory(10); // Initialize inventory with 10 shelves
        this.coinList = new ArrayList<>();
        System.out.println("Initialized Vending Machine in Idle State");
    } 

    // State management
    public void setState(VendingMachineState newState){
        this.currMachineState = newState;
    }
    public VendingMachineState getCurrMachineState(){
        return currMachineState;
    }
    public void setPaymentStrategy(PaymentStrategy paymentStrategy){
        this.paymentStrategy = paymentStrategy;
    }
   
    // Events (delegated to current state)
    public void insertCoin(Coin coin){
        currMachineState.insertCoin(this, coin);
    }
    public void selectItem(int codeNumber){
        currMachineState.selectItem(this, codeNumber);
    }
    public void dispense(){
        currMachineState.dispenseItem(this);
    }

    public boolean processPayment(int amount){
        if(paymentStrategy == null){
            throw new IllegalStateException("Payment strategy not set");
        }
        return paymentStrategy.processPayment(amount);
    }
    // Domain operations (used by states)
    public Inventory getInventory(){
        return inventory;
    }

    public List<Coin> getCoinList(){
        return coinList;
    }

    public void addCoin(Coin coin){
        coinList.add(coin);
    }

    public int getBalance(){
        int balance = 0;
        for(Coin coin : coinList){
            balance += coin.getValue();
        }
        return balance;
    }

    public void clearBalance(){
        coinList.clear();
    }

    public int getSelectedItemCode(){
        return selectedItemCode;
    }

    public void setSelectedItemCode(int codeNumber){
        this.selectedItemCode = codeNumber;
    }

    public void resetSelection(){
        this.selectedItemCode = 0;
    }
}

File: Python/A_Project/E_Vending_Machine/VendingMachineStates/vending_machine_context.py
from CommonEnums.coin import Coin
from PaymentStrategy.payment_strategy import PaymentStrategy
from UtilityClasses.inventory import Inventory
from VendingMachineStates.ConcreteStates.idle_state import IdleState


class VendingMachineContext:
    def __init__(self):
        self._curr_machine_state = IdleState()
        self._payment_strategy = None
        self._inventory = Inventory(10)
        self._coin_list = []
        self._selected_item_code = 0
        print("Initialized Vending Machine in Idle State")

    def set_state(self, new_state):
        self._curr_machine_state = new_state

    def get_curr_machine_state(self):
        return self._curr_machine_state

    def set_payment_strategy(self, payment_strategy: PaymentStrategy):
        self._payment_strategy = payment_strategy

    def insert_coin(self, coin: Coin):
        self._curr_machine_state.insert_coin(self, coin)

    def select_item(self, code_number: int):
        self._curr_machine_state.select_item(self, code_number)

    def dispense(self):
        self._curr_machine_state.dispense_item(self)

    def process_payment(self, amount: int) -> bool:
        if self._payment_strategy is None:
            raise RuntimeError("Payment strategy not set")
        return self._payment_strategy.process_payment(amount)

    def get_inventory(self) -> Inventory:
        return self._inventory

    def get_coin_list(self):
        return self._coin_list

    def add_coin(self, coin: Coin):
        self._coin_list.append(coin)

    def get_balance(self) -> int:
        return sum(coin.get_value() for coin in self._coin_list)

    def clear_balance(self):
        self._coin_list.clear()

    def refund_payment(self):
        balance = self.get_balance()
        if balance > 0:
            print(f"Refunding amount: {balance}")
        elif self._payment_strategy is not None:
            print("Voiding card payment.")
        self.clear_balance()

    def get_selected_item_code(self) -> int:
        return self._selected_item_code

    def set_selected_item_code(self, code_number: int):
        self._selected_item_code = code_number

    def reset_selection(self):
        self._selected_item_code = 0
